boot/freeldr/freeldr/arch/realmode/helpers_all.inc

```diff
@@ -0,0 +1,81 @@
+/*
+ * PROJECT:     FreeLoader
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     Real mode helper code for all architectures
+ * COPYRIGHT:   Copyright 2024 Daniel Victor <ilauncherdeveloper@gmail.com>
+ */
+
+Relocator16SS:
+    .word 0
+Relocator16EAX:
+    .long 0
+Relocator16ESP:
+    .long 0
+
+cs_writestr:
+    push ax
+    push ds
+    mov ax, cs
+    mov ds, ax
+
+    call writestr
+    pop ds
+    pop ax
+    ret
+
+GetCurrentBaseAddress:
+    /* Clear necessary registers */
+    xor eax, eax
+    xor ebx, ebx
+    xor ecx, ecx
+
+    /* Call the sub function to get the base address */
+    call .SubCall
+
+.SubCall:
+    /* Pop the return address into bx */
+    pop bx
+
+    /* Calculate the base address by adjusting the offset */
+    sub bx, offset .SubCall
+
+    /* Load the current code segment into eax */
+    mov ax, cs
+
+    /* Shift the segment address to get its physical address (real mode calculation) */
+    shl eax, 4
+
+    /* Add the physical segment base to the relative base address */
+    add eax, ebx
+
+    /* Store the calculated base address (low 32 bits) */
+    mov dword ptr ds:[BSS_CurrentBaseAddress], eax
+
+    /* Clear the high 32 bits (not used in 32-bit environments) */
+    mov dword ptr ds:[BSS_CurrentBaseAddress + 4], ecx
+
+    /* Save the return address offset */
+    mov ecx, ebx
+
+    /* Convert the base address to segment for far jump */
+    mov ebx, eax
+    shr ebx, 4
+
+    /* Calculate the address of the far jump instruction */
+    add eax, offset .FarJump
```

Comment:
I don't fully understand yet why we need all this far-jump stuff here for getting the current base address.
Could you explain it to me why?

Also, when code does `call GetCurrentBaseAddress`, are segments (CS, DS....) modified once this call returns?

Reply:
I added the far-jump stuff in this function because i still have the return address stored in `ebx` and then passed to `ecx`.